sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/aio/_client.py

```diff
@@ -0,0 +1,1091 @@
+
+                if not self._is_initial_connected:
+                    self._is_initial_connected = True
+                    if self._auto_rejoin_groups:
+                        for group_name, group in self._group_map.items():
+                            if group.is_joined:
+                                asyncio.create_task(self._rejoin_group(group_name))
+
+                    self._call_back(
+                        CallbackType.CONNECTED,
+                        OnConnectedArgs(connection_id=message.connection_id, user_id=message.user_id),
+                    )
+
+            async def handle_disconnected_message(message: DisconnectedMessage):
+                self._last_disconnected_message = message
+
+            async def handle_group_data_message(message: GroupDataMessage):
+                if message.sequence_id:
+                    if not self._sequence_id.try_update(message.sequence_id):
+                        # // drop duplicated message
+                        return
+
+                self._call_back(
+                    CallbackType.GROUP_MESSAGE,
+                    OnGroupDataMessageArgs(
+                        data_type=message.data_type,
+                        data=message.data,
+                        group=message.group,
+                        from_user_id=message.from_user_id,
+                        sequence_id=message.sequence_id,
+                    ),
+                )
+
+            async def handle_server_data_message(message: ServerDataMessage):
+                if message.sequence_id:
+                    if not self._sequence_id.try_update(message.sequence_id):
+                        # // drop duplicated message
+                        return
+
+                self._call_back(
+                    CallbackType.SERVER_MESSAGE,
+                    OnServerDataMessageArgs(
+                        data_type=message.data_type,
+                        data=message.data,
+                        sequence_id=message.sequence_id,
+                    ),
+                )
+
+            parsed_message = self._protocol.parse_messages(data)
+            if parsed_message is None:
+                #  None means the message is not recognized.
+                return
+            if parsed_message.kind == "connected":
+                await handle_connected_message(parsed_message)
+            elif parsed_message.kind == "disconnected":
+                await handle_disconnected_message(parsed_message)
+            elif parsed_message.kind == "ack":
+                await handle_ack_message(parsed_message)
+            elif parsed_message.kind == "groupData":
+                await handle_group_data_message(parsed_message)
+            elif parsed_message.kind == "serverData":
+                await handle_server_data_message(parsed_message)
+            else:
+                _LOGGER.warning("unknown message type: %s", parsed_message.kind)
+
+        async def _listen():
```

Comment:
This functionality here should be inside the `WebSocketAppAsync` class. So it should run receive in a background task and then fire the registered callbacks (`on_close`, `on_message`, similar to the sync set up [here](https://github.com/Azure/azure-sdk-for-python/blob/37eb477bd10be094b7692b1361106e14659e9410/sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/_client.py#L811)). This way we can keep the websocket implementation separate from the webpubsub implementation.

Reply:
I see.